src/Bicep.Core/TypeSystem/DeployTimeConstantDirectViolationVisitor.cs

```diff
@@ -16,18 +17,35 @@ public DeployTimeConstantDirectViolationVisitor(SyntaxBase deployTimeConstantCon
 
         public override void VisitArrayAccessSyntax(ArrayAccessSyntax syntax)
         {
-            if (syntax.IndexExpression is StringSyntax stringSyntax &&
-                this.ResourceTypeResolver.TryResolveResourceOrModuleSymbolAndBodyType(syntax.BaseExpression) is ({ } accessedSymbol, { } accessedBodyType))
+            if (this.ResourceTypeResolver.TryResolveResourceOrModuleSymbolAndBodyType(syntax.BaseExpression) is ({ } accessedSymbol, { } accessedBodyType))
             {
-                if (stringSyntax.TryGetLiteralValue() is { } propertyName)
+                var indexExprTypeInfo = SemanticModel.GetTypeInfo(syntax.IndexExpression);
+                if (indexExprTypeInfo is StringLiteralType { RawStringValue: var propertyName })
                 {
                     // Validate property access via string literal index (myResource['sku']).
                     this.FlagIfPropertyNotReadableAtDeployTime(syntax, propertyName, accessedSymbol, accessedBodyType);
                 }
+                else if (indexExprTypeInfo is UnionType { Members: var indexUnionMembers } )
+                {
+                    var unionMemberTypes = indexUnionMembers.Select(m => m.Type).ToList();
+                    if (unionMemberTypes.All(t => t is StringLiteralType))
+                    {
+                        foreach (var unionMemberType in unionMemberTypes.Cast<StringLiteralType>())
+                        {
+                            if (this.FlagIfPropertyNotReadableAtDeployTime(syntax, unionMemberType.RawStringValue, accessedSymbol, accessedBodyType))
+                            {
+                                break;
+                            }
```

Comment:
Why only flag the first DTC violation?

Reply:
The BCP182 diagnostic does not currently have the property name in the message. It only lists the properties that are deploy time accessible. Without this condition, two diagnostics appear with the exact same message in the case of something like `foo[cond ? 'extendedLocation' : 'properties']`. 

Alternatives could be to list out all the properties that are flagged in a single diagnostic, or provide a single property name in the message and flag each union member.

Reply:
I would go with the second option (accept a single property name in the diagnostic builder and flag each union member).

Do you know when a union would be allowed? I wouldn't think `<resource>['apiVersion' | 'id' | 'type']` would count as a deploy time constant, even if each individual member of the union would be a deploy-time constant property name.

Reply:
I'm not sure. It looks to depend on the expression and how the `GetTypeInfo` method resolves it and how/when these evaluations are done. To validate this, I'd need to run an ARM template validation or manually test a deployment?

Reply:
Hit some emitter exceptions while manually testing cases like
```bicep
param condParam bool = false

resource foos 'Microsoft.Storage/storageAccounts@2022-09-01' = [for i in range(0, 2): {
  name: '${i}foo${uniqueString(resourceGroup().id)}'
  location: 'westus'
  sku: {
    name: 'Standard_LRS'
  }
  kind: 'StorageV2'
}]

var test = [for i in range(0, 2): foos[i][condParam ? 'id' : 'apiVersion']]

resource dtcTests 'Microsoft.Storage/storageAccounts@2022-09-01' = [for i in range(0, 2): {
  name: 'dtc${test[i]}'
  location: 'westus'
  sku: {
    name: 'Standard_LRS'
  }
  kind: 'StorageV2'
}]
```

Due to the emitter issues, this may need to be revisited another time. It feels like this should be valid. `test` would always have values of ids or apiVersions which are supposed to be deploy time constants. The usage of `test` is also in a deploy time context. If there is a union resolved in the property accessor, they need to be string literals which are resolved at compile time. Parameters for instance are not resolved as string literals but as a plain string type.